junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/NamespaceAwareStore.java

```diff
@@ -66,9 +70,14 @@ public <K, V> V getOrComputeIfAbsent(K key, Function<K, V> defaultCreator, Class
 			() -> this.valuesStore.getOrComputeIfAbsent(this.namespace, key, defaultCreator, requiredType));
 	}
 
+	@SuppressWarnings("deprecation")
 	@Override
 	public void put(Object key, Object value) {
 		Preconditions.notNull(key, "key must not be null");
+
+		if (value instanceof CloseableResource && !(value instanceof AutoCloseable)) {
+			LOGGER.warn(() -> "The object implements CloseableResource but not AutoCloseable: " + value.getClass());
```

Comment:
I think we should defer this warning until we close the store (see previous comment).